File: mydriver/lib/features/driver/logic/driver_controller.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import '/core/models/booking.dart';

class DriverController with ChangeNotifier {
  bool _isLoading = false;
  String? _errorMessage;
  List<Booking> _orders = [];

  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  List<Booking> get orders => _orders;

  // Fetch orders for the driver
  Future<void> fetchOrders(String driverId) async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    try {
      final response = await http.get(
        Uri.parse('http://localhost:3000/api/driver/orders?driverId=$driverId'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        _orders = data.map((item) => Booking.fromJson(item)).toList();
        _isLoading = false;
        notifyListeners();
      } else {
        _errorMessage =
            jsonDecode(response.body)['message'] ?? 'Failed to fetch orders';
        _isLoading = false;
        notifyListeners();
      }
    } catch (e) {
      _errorMessage = 'Error connecting to the server';
      _isLoading = false;
      notifyListeners();
    }
  }

  // Accept an order
  Future<bool> acceptOrder(String orderId) async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    try {
      final response = await http.post(
        Uri.parse('http://localhost:3000/api/driver/accept-order'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'orderId': orderId}),
      );

      if (response.statusCode == 200) {
        _orders.firstWhere((order) => order.id == orderId).status = 'accepted';
        _isLoading = false;
        notifyListeners();
        return true;
      } else {
        _errorMessage =
            jsonDecode(response.body)['message'] ?? 'Failed to accept order';
        _isLoading = false;
        notifyListeners();
        return false;
      }
    } catch (e) {
      _errorMessage = 'Error connecting to the server';
      _isLoading = false;
      notifyListeners();
      return false;
    }
  }
}

File: mydriver-backend/src/controllers/driverController.js
const Booking = require('../models/booking');

exports.getOrders = async (req, res) => {
  try {
    const { driverId } = req.query;
    if (!driverId) {
      return res.status(400).json({ message: 'Driver ID is required' });
    }

    const orders = await Booking.find({ driverId, status: { $in: ['pending', 'accepted'] } });
    res.status(200).json(orders);
  } catch (error) {
    res.status(500).json({ message: 'Error fetching orders', error: error.message });
  }
};

exports.acceptOrder = async (req, res) => {
  try {
    const { orderId } = req.body;
    if (!orderId) {
      return res.status(400).json({ message: 'Order ID is required' });
    }

    const booking = await Booking.findById(orderId);
    if (!booking) {
      return res.status(404).json({ message: 'Order not found' });
    }

    booking.status = 'accepted';
    await booking.save();

    res.status(200).json({ message: 'Order accepted successfully' });
  } catch (error) {
    res.status(500).json({ message: 'Error accepting order', error: error.message });
  }
};